core/controllers/library.py

```diff
@@ -81,6 +81,14 @@ def get(self):
         else:
             page_mode = feconf.LIBRARY_PAGE_MODE_INDEX
 
+        is_redirect = True
+        if self.request.referer and (
+                not self.request.referer.endswith('/library')):
+            is_redirect = False
+
+        if self.user_id is None and not is_redirect:
+            self.redirect('/library')
```

Comment:
IIRC you need to do a "return" after calling self.redirect() -- it doesn't terminate immediately.

Also I am a little confused about what is going on. IIUC all of the code added here is basically equivalent to:

```
        if self.user_id is None and self.request.referer and not self.request.referer.endswith('/library'):
            self.redirect('/library')
```

and I don't really understand the logic behind it. Could you please explain?

Thanks!

Reply:
Hi @seanlip,

The issue occurs because the user_is_logged_in parameter does not get updated when the library.html template is rendered. User needs to reload the page to see the changes. 

In order to fix this, it was necessary to redirect to library once again if user_id is still None after a login. This fixes the issue when the user has logged in as the page is refreshed virtually.

This fix was creating an issue in the case where there is no user logged in and there were multiple redirects. So, to avoid that, if the page is referred from /library once, there has been one redirect already and we need to stop.

Do let me know if anything else is not clear, Thanks!

Reply:
Wow, that's a lot of important context! It would have been good to see that represented as a (long) comment in the PR.

It sounds like you're not fixing the root cause, though. Can you explain / figure out _why_ the user_is_logged_in parameter does not get updated when the library.html template is rendered? We should be fixing that instead.

Reply:
Hi @seanlip, 
I tried to dig a little deeper into the issue. When the user clicks on library link, the values should get updated ideally [here](https://github.com/oppia/oppia/blob/develop/core/controllers/base.py#L306) but none of the values are updated. 

I tried to google about it and found a link: https://stackoverflow.com/questions/19117061/web-application-is-not-reloaded-automatically-since-gae-migration-from-python-2. This links mentions about eventual and strong consistency. I am not a lot familiar with this and I am not sure whether this is causing the issue. May you please help me here, Thanks!

Reply:
Hi @bansalnitish, thanks for the investigation. Could you please clarify the user flow in the original issue a bit more with screenshots? I realized while trying to write a reply that I am not 100% certain how to reproduce this.

In particular, when you say "login", what does this involve? Is this different from signup? What page do I land on before I click the library button?

Reply:
Hi @seanlip, sorry for the confusion. I have now added a gif showing the complete process. PTAL, Thanks!

Reply:
Thanks for the clarification @bansalnitish!

Here is what I suggest. If you follow the repro instructions you wrote, but change the last step to "navigate to the About Oppia page", everything works fine.

So, focus on figuring out what's different between the Library page and the "About Oppia" page. Why does one show the user profile and not the other? If it is indeed the backend handler you pointed to, why doesn't that affect the About page? (To tackle this you need to do enough digging such that you have a good understanding of the end-to-end flow that causes the user profile to be shown.)

Reply:
Thanks @seanlip, I'll check and let you know after further investigation.